### PR TITLE
Return soft failures in most PIO inquire functions

### PR DESCRIPTION
**Pull request by dqwu**
A failure during a PIO inquire function should not be considered
fatal in most cases. 

Certain inquire function failures are fatal, though. For example,
inquiring the name/size of a type (PIOc_inq_type()).
    
For a soft (non-fatal) failure, the error code is broadcast to
all processes and is directly returned to the user.
    
We no longer call pio_err() on a soft failure, as it aborts the
program if the error handler is set to PIO_INTERNAL_ERROR.

This change allows an ES3M case to continue if a failed PIO
inquiry is actually expected.

### CODE REVIEW

File: src/clib/pio_nc.c [hunk 1]
```diff
@@ -119,13 +119,13 @@ int PIOc_inq(int ncid, int *ndimsp, int *nvarsp, int *ngattsp, int *unlimdimidp)
 
     /* A failure to inquire is not fatal */
     mpierr = MPI_Bcast(&ierr, 1, MPI_INT, ios->ioroot, ios->my_comm);
-    if(mpierr != MPI_SUCCESS){
+    if (mpierr != MPI_SUCCESS)
```
Comment (jayeshkrishna) [inline src/clib/pio_nc.c]: This is a code formatting/style change and needs to go separately. IMO, it is always a good idea to add braces for statements (or statement) within an if condition.

File: src/clib/pio_nc.c [hunk 2]
```diff
@@ -315,13 +315,13 @@ int PIOc_inq_unlimdims(int ncid, int *nunlimdimsp, int *unlimdimidsp)
...
```
Comment (jayeshkrishna) [inline src/clib/pio_nc.c]: code formatting change, needs to go in a separate PR

File: src/clib/pio_nc.c [hunk 3]
```diff
@@ -394,9 +394,13 @@ int PIOc_inq_type(int ncid, nc_type xtype, char *name, PIO_Offset *sizep)
         LOG((2, "PIOc_inq_type netcdf call returned %d", ierr));
     }
 
-    /* Failure to inquire is fatal */
-    ierr = check_netcdf(NULL, file, ierr, __FILE__, __LINE__);
-    if(ierr != PIO_NOERR){
+    /* A failure to inquire is not fatal */
```
Comment (jayeshkrishna) [inline src/clib/pio_nc.c]: I had determined before that certain inq function failures were fatal. For example, inquiring the name/size of a type (PIOc_inq_type() ) .
So I think in this PR we only need to modify the cases where the comments specifically mention "Failure to inquire is not fatal"
There are several cases below that need to be similarly reverted back.

**Comment by dqwu** [comment]
@jayeshkrishna 
I have rebased the feature branch, and updated information of this PR.

**Comment by jayeshkrishna** [comment]
One more thing to consider here is whether we want to sync "soft errors" across all processes. Since applications could perform actions based on the return values from these functions, it might make sense to sync these return values. For example,
1) Inquire if a variable is present in the file
2) if var not present call func_var_NOT_found()
3) if var present call func_var_found()

Let me know your thoughts.

**Comment by jayeshkrishna** [comment]
Returning errors from the underlying I/O library, assumes that pio error codes map to these error codes (a netcdf error code is a pio error code). Although this is the case now, we might need to rethink this strategy in future.

File: src/clib/pio_nc.c [hunk 4]
```diff
@@ -1340,8 +1340,8 @@ int PIOc_inq_attname(int ncid, int varid, int attnum, char *name)
     }
 
     if(ierr != PIO_NOERR){
-        LOG((1, "nc*_inq failed, ierr = %d", ierr));
-        return pio_err(ios, file, ierr, __FILE__, __LINE__);
+        LOG((1, "nc*_inq_attname failed, ierr = %d", ierr));
```
Comment (jayeshkrishna) [inline src/clib/pio_nc.c]: Do you think this case should be a fatal error? The user is already specifying the attnum and inquiring the corresponding attribute name.

Comment (jayeshkrishna) [reply src/clib/pio_nc.c]: The old code considered it as a "soft error" but maybe this case should be fatal

**Comment by jayeshkrishna** [comment]
> One more thing to consider here is whether we want to sync "soft errors" across all processes. Since applications could perform actions based on the return values from these functions, it might make sense to sync these return values. For example,
> 
>     1. Inquire if a variable is present in the file
> 
>     2. if var not present call func_var_NOT_found()
> 
>     3. if var present call func_var_found()
> 
> 
> Let me know your thoughts.

After looking at the source, for the cases that are considered "non fatal" or "soft errors" we broadcast error code from io rank 0, so the error codes are synced globally.

File: src/clib/pio_nc.c (hunk 4 above)
Comment (dqwu) [reply src/clib/pio_nc.c]: I will add the change to convert nc_inq_attname from soft to
a fatal error as a separate commit.

**Comment by dqwu** [comment]
@jayeshkrishna 
I have pushed a new commit to add the change to convert
nc*_inq_attname from soft to a fatal error.
